### PR TITLE
TR-71: Add TS to derive and derive/work-timing

### PR DESCRIPTION
**Pull request by shoumaw**
https://bitovi.atlassian.net/browse/TR-71
Added type safety to

1. derive
2. derive/work-timing
3. shared/issue-data/date-data
Plus some moving up of types to the shared types file

Add tests to 

1. derive
2. derive/work-timing

These are needed to unblock adding type safety to `rollup`.

### CODE REVIEW

File: public/jira/derived/derive.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import {
+  DerivedWorkStatus,
+  getWorkStatus,
+  WorkStatusConfig,
+} from "./work-status/work-status.js";
+import {
+  normalizeIssue,
+  NormalizeIssueConfig,
+} from "../normalized/normalize.js";
+import { JiraIssue, NormalizedIssue } from "../shared/types.js";
```
Comment (DavidNic11) [inline public/jira/derived/derive.ts:15]: should these imports be from TypeScript files?

edit: same with the other added files

File: public/jira/derived/derive.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * @param {NormalizedIssue} normalizedIssue
+ * @return {DerivedIssue}
+ */
+export function deriveIssue(
+  issue: NormalizedIssue,
+  options: Partial<WorkStatusConfig & WorkTimingConfig> & {
+    uncertaintyWeight?: number;
+  } = {}
+): DerivedIssue {
+  const derivedTiming = deriveWorkTiming(issue, options);
```
Comment (DavidNic11) [inline public/jira/derived/derive.ts:38]: `options` isn't being used, given the pattern in the other files I think these (`deriveWorkTiming` and `getWorkStatus`) should be defaulted on options

File: public/jira/derived/work-timing/work-timing.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  sampleExtraPoints,
+} from "../../../shared/confidence.js";
+import {
+  DueData,
+  getStartDateAndDueDataFromFieldsOrSprints,
+  getStartDateAndDueDataFromSprints,
+  StartData,
+} from "../../../shared/issue-data/date-data.js";
+import {
+  DefaultsToConfig,
```
Comment (DavidNic11) [inline public/jira/derived/work-timing/work-timing.ts:13]: Nice. TY for this

File: public/jira/derived/work-timing/work-timing.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  getConfidenceDefault,
+  getStoryPointsDefault,
+};
+
+export type WorkTimingConfig = DefaultsToConfig<typeof defaults>;
+
+/**
+ *
+ * @param {import("../../shared/types.js").NormalizedIssue} normalizedIssue
+ * @param {*} param1
```
Comment (DavidNic11) [inline public/jira/derived/work-timing/work-timing.ts:74]: `param1`?

File: public/jira/derived/work-timing/work-timing.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+ * @returns {DerivedWorkTiming}
+ */
+export function deriveWorkTiming(
+  normalizedIssue: NormalizedIssue,
+  {
+    getConfidence = getConfidenceDefault,
+    getStoryPoints = getStoryPointsDefault,
+    uncertaintyWeight = 80,
+  }: Partial<WorkTimingConfig> & { uncertaintyWeight?: number } = {}
+) {
```
Comment (DavidNic11) [inline public/jira/derived/work-timing/work-timing.ts:84]: do we need a return type?

edit: After looking at the previous and rest of the changes, we should probably take a stance on whether we add return types or not. Generally, I prefer to explicitly type return types on functions that are exported unless there's a reason not to... What are your thoughts?

File: public/jira/derived/work-timing/work-timing.ts (hunk 4 above)
Comment (shoumaw) [reply public/jira/derived/work-timing/work-timing.ts:74]: Oops, this is how it existed in the js version and I forgot to fix it.

File: public/jira/derived/work-timing/work-timing.ts (hunk 5 above)
Comment (shoumaw) [reply public/jira/derived/work-timing/work-timing.ts:84]: I added explicit return types for all functions

File: public/jira/derived/derive.ts (hunk 2 above)
Comment (shoumaw) [reply public/jira/derived/derive.ts:35]: `options` is intended for the callbacks being used under `deriveWorkTiming` and `getWorkStatus` where it gets a default value there.

Its not meant for passing an implementation to `deriveWorkTiming` and `getWorkStatus`.

Does that make sense?

**Review by DavidNic11 — APPROVED** [review]

**Pull request merged**